### ISSUE TITLE
dotspacemacs-auto-resume-layouts interferes with hooks

### ISSUE THREAD

**Issue by mkomitee** [syl20bnr/spacemacs#3881]
After enabling `dotspacemacs-auto-resume-layouts` (running on the develop branch) I've noticed that all buffers are loaded before my `dotspacemacs/user-config` runs, which means any hooks I add in there don't fire when the buffers load. 

Is this expected / is that the wrong place to configure my hooks?

For now, I've been moving these types of hooks configuration into `custom-set-variables`, which I'd rather not have to do because I can't comment them or format them.

**Comment by mkomitee** [comment]
For example, the following elisp:

```
(add-hook 'prog-mode-hook 'turn-on-fci-mode)
```

Doesn't cause `fci-mode` to be enabled in any `prog-mode` buffers opened by the auto-resumed layout. Only manually opened files will have `fci-mode` enabled.

**Comment by bmag** [comment]
Hey @mkomitee, sorry for late response. I've tried to reproduce the bug today but was not successful.

This is the configuration I used:
#### System Info
- OS: gnu/linux
- Emacs: 24.5.1
- Spacemacs: 0.105.9
- Spacemacs branch: develop (rev. 72e1413)
- Distribution: spacemacs
- Editing style: hybrid
- Completion: helm
- Layers:
```elisp
(spacemacs-helm
 (auto-completion :variables auto-completion-enable-snippets-in-popup t auto-completion-enable-help-tooltip t company-tooltip-align-annotations t)
 emacs-lisp git markdown org
 (shell :variables shell-default-height 30 shell-default-position 'bottom)
 syntax-checking version-control python eyebrowse themes-megapack c-c++ ibuffer)
```
And I've changed the value of `dotspacemacs-auto-resume-layouts` to `t` in `dotspacemacs/init`.

These are the steps I've tried:
- start spacemacs
- open new layout (`SPC l 2`) and name it "test"
- open a python file
- restart spacemacs

Observed behavior: spacemacs start normally and shows the home buffer, after ~1 second spacemacs loads my layouts from the previous session. I switch to "test" layout (`SPC l 2`), switch to the python file (`SPC l b`), and see that `fci-mode` is enabled correctly in the python file.

Are you still experiencing the bug?

**Comment by mkomitee** [comment]
I'm still having this problem. I'll try to get back to you with more information.

**Comment by bmag** [comment]
Can you provide a step-by-step recipe to reproduce the problem, and your system info? You can use [this template](https://github.com/syl20bnr/spacemacs/wiki/Debugging#how-to-make-a-great-bug-report).

**Comment by syl20bnr** [comment]
We should tackle this bug, seems pretty nasty.

**Comment by ntcong** [comment]
I found that including vim-empty-lines is the problem. With vim-empty-lines, hooks will not be execute on opened buffer, it will still work on new buffer, but will not work again after restart emacs

**Comment by ntcong** [comment]
It seems that vim-empty-lines is not the only layer causing problem, but with this minimal layer remove vim-empty-lines will help

```
  dotspacemacs-configuration-layers
  '(
    spacemacs-ivy
    spacemacs-layouts
    vim-empty-lines
    python
  )
```

This is the hook I'm using

```
(defun dotspacemacs/user-config ()
  "This is were you can ultimately override default Spacemacs configuration.
This function is called at the very end of Spacemacs initialization."
  (add-hook 'python-mode-hook 'highlight-indentation-mode)
  (add-hook 'go-mode-hook 'highlight-indentation-mode)
)
```

**Comment by aaronjensen** [comment]
This is a very general issue that has bitten me several times. The problem is that `dotspacemacs-auto-resume-layouts` reopens buffers after `user-init`, but before `user-config`. As such, anything you do in user-config that needs to be in place prior to a buffer being open (like setting `web-mode-markup-indent-offset`) before opening a web-mode file will not take effect in the buffers that are opened.

If you move your configuration to `user-init` the problem goes away.

In short, the current run order is:

1. user-init
2. open previously opened buffers
3. user-config

and it should be

1. user-init
2. user-config
3. open previously opened buffers

**Comment by cpaulik** [comment]
Could some of you please test https://github.com/syl20bnr/spacemacs/pull/5962 since I'm not sure that the fix is so straightforward.

**Comment by aaronjensen** [comment]
@cpaulik tested and confirmed it fixed my scenario. Thank you!

**Comment by d12frosted** [comment]
Fixed with released of Spacemacs v0.200. Let us know if you still have any problems with this issue.

**Comment by bmag** [comment]
@zzjjzzgggg there is, should be named "persp-something-hook" or "persp-something-functions" (persp-mode provides the layouts feature)
